org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperJavaImpl.java

```diff
@@ -352,49 +353,60 @@ protected Void handle(URI uri, IStorage storage,
 	/* @NonNull */
 	@Override
 	public Iterable<Pair<IStorage, IProject>> getStorages(/* @NonNull */ URI uri) {
-		List<Pair<IStorage, IProject>> result = newArrayListWithCapacity(1);
-		List<PackageFragmentRootData> packageFragmentRootDatas;
+		// Pessimistic copy of the cachedPackageFragmentRootData
+		Collection<PackageFragmentRootData> packageFragmentRootDatas;
 		synchronized(cachedPackageFragmentRootData) {
-			packageFragmentRootDatas = newArrayList(cachedPackageFragmentRootData.values());
-		}
-		Iterator<PackageFragmentRootData> iterator = packageFragmentRootDatas.iterator();
-		while (iterator.hasNext()) {
-			PackageFragmentRootData data = iterator.next();
-			if (data.exists()) {
-				if (data.uriPrefix == null || uri.toString().startsWith(data.uriPrefix.toString())) {
-					IStorage storage = data.uri2Storage.get(uri);
-					if (storage != null) {
+			packageFragmentRootDatas = cachedPackageFragmentRootData.values();
+		}
+		
+		/* 
+		 * We iterate all known package fragment roots and put them into two buckets if matching:
+		 * - uriPrefix == null || uri startsWith uriPrefix
+		 *   => regularMatch
+		 * 
+		 * - uri.isArchive && archiveURI.isFile || isPlatformResource && uriPrefix is not null && no other matches && 
+		 *   => archiveMatch
+		 */
+		List<Pair<IStorage, IProject>> regularMatches = new ArrayList<Pair<IStorage,IProject>>(1);
+		List<Pair<IStorage, IProject>> archiveMatches = null;
+		IPath archivePath = null;
+		if (uri.isArchive()) {
+			String authority = uri.authority();
+			URI archiveURI = URI.createURI(authority.substring(0, authority.length() - 1));
+			if (archiveURI.isFile() || archiveURI.isPlatformResource()) {
+				archivePath = new Path(archiveURI.isPlatformResource()? archiveURI.toPlatformString(true): archiveURI.toFileString());
+				archiveMatches = new ArrayList<Pair<IStorage,IProject>>(1);
+			}
```

Comment:
is here an else case? nested archives?

Reply:
As far as I understood from the previous impl, there was no else case. Content of nested archives is usually not on the classpath of the project either. I think it's fine as is.